app/lib/shared/utils.dart

```diff
@@ -361,3 +362,38 @@ bool fixedTimeEquals(String a, String b) {
   }
   return result == 0;
 }
+
+extension StreamBoundedForEach<T> on Stream<T> {
+  /// Executes items using [eachFn] with maximum [concurrency].
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T) eachFn,
+  ) async {
+    final pool = Pool(concurrency);
+    final futures = Queue<Future>();
+    // keeps the size of the futures limited, allows some difference in the run times
+    final maxQueueSize = concurrency * 2;
+    try {
+      await for (final item in this) {
+        final f = pool.withResource(() => eachFn(item));
+        futures.add(f);
+        while (futures.length > maxQueueSize) {
+          await futures.removeFirst();
+        }
+      }
+      await Future.wait(futures);
+    } finally {
+      await pool.close();
+    }
+  }
```

Comment:
```suggestion
    Future<void> boundedForEach(int concurrency, FutureOr<void> Function(T) each) async {
    try {
      var N = 0;
      var Object? err;
      var StackTrace? stackTrace;
      var blocked = Completer()..complete();
      await for (final item in this) {
        scheduleMicroTask(() async {
          try {
            await each(item);
          } catch (e, st) {
            if (err == null) {
              (err, stackTrace) = (e, st);
            }
          } finally {
            N--;
            if (!blocked.isCompleted) {
              blocked.complete();
            }
          }
        });
        N++;
        if (N >= concurrency) {
          blocked = Completer();
          await blocked.future;
        }
      }
    } finally {
      while (N > 0) {
        blocked = Completer();
        await blocked.future;
      }
    }
  }
```

Reply:
I think this will schedule all of the async task upfront (e.g. for 1 million entries it may create 1 million microtask before the first completes). The benefit of the queue + blocking is that it won't have more than the bound condition at once (assuming that the original iteration pauses when the async processing is being blocked).